lamindb/core/history/_trigger_installer.py

```diff
@@ -0,0 +1,229 @@
+        FOR EACH ROW EXECUTE PROCEDURE {function_name}();
+        """)
+
+    def _install_insert_trigger(self, table: str, cursor: CursorWrapper) -> str:
+        return self._install_trigger(
+            table=table,
+            history_event_type=HistoryEventTypes.INSERT,
+            function_body_sql="""
+        INSERT INTO lamindb_history
+        (id, migration_history_id, table_name, record_id, record_data, event_type, created_at)
+        VALUES
+        (gen_random_uuid(), latest_migration, %s, NEW.id, row_to_json(NEW), %s, now());
+        """,
+            cursor=cursor,
+        )
+
+    def _install_update_trigger(self, table: str, cursor: CursorWrapper):
+        return self._install_trigger(
+            table=table,
+            history_event_type=HistoryEventTypes.UPDATE,
+            function_body_sql="""
+        INSERT INTO lamindb_history
+        (id, migration_history_id, table_name, record_id, record_data, event_type, created_at)
+        VALUES
+        (gen_random_uuid(), latest_migration, %s, NEW.id, row_to_json(NEW), %s, now());
+        """,
+            cursor=cursor,
+        )
+
+    def _install_delete_trigger(self, table: str, cursor: CursorWrapper):
+        return self._install_trigger(
+            table=table,
+            history_event_type=HistoryEventTypes.DELETE,
+            function_body_sql="""
+        INSERT INTO lamindb_history
+        (id, migration_history_id, table_name, record_id, record_data, event_type, created_at)
+        VALUES
+        (gen_random_uuid(), latest_migration, %s, OLD.id, NULL, %s, now());
+        """,
+            cursor=cursor,
+        )
+
+    def install_triggers(self, table: str):
+        if not self.VALID_TABLE_NAME_REGEX.match(table):
+            raise ValueError(
+                f"Table name '{table}' doesn't look like a valid PostgreSQL table name"
+            )
+
+        cursor = self.connection.cursor()
+
+        # TODO: should we be extracting the primary key from info schema as well?
+        # The assumption that all tables are keyed off 'id' seems brittle.
+
+        # TODO: migration_history_id is only capturing the latest migration from
+        # the `lamindb` app, and isn't taking bionty, etc into account.
+
+        self._install_insert_trigger(table, cursor)
+        self._install_update_trigger(table, cursor)
+        self._install_delete_trigger(table, cursor)
+
+        print("Trigger functions installed")
+
+    def get_tables_with_installed_triggers(self) -> set[str]:
+        cursor = self.connection.cursor()
+
+        cursor.execute("""
+SELECT
+    DISTINCT event_object_table
+FROM information_schema.triggers
+WHERE trigger_name like 'lamindb_history_%';
+""")
+        tables = set()
+
+        rows = cursor.fetchall()
+
+        for row in rows:
+            table_name = row[0]
+            tables.add(table_name)
+
+        return tables
+
+
+class SQLiteHistoryRecordingTriggerInstaller(HistoryRecordingTriggerInstaller):
```

Comment:
Just to be sure: we'll be able to add an sqlite version of the triggers in the future, right? I would ask AI to translate it from Postgres dialect to SQLite.

I'm also thinking to potentially create a test group only for the history tracking because it's entirely orthogonal to everything else.

It's the first feature that will truly differ across SQLite and Postgres and it's the first feature that operates on the database level. Everything else in lamindb is just python logic on top of Django and fsspec.

Reply:
> Just to be sure: we'll be able to add an sqlite version of the triggers in the future, right?

Yup! The SQLite triggers won't look exactly the same, but they will look _very_ similar.

> I'm also thinking to potentially create a test group only for the history tracking because it's entirely orthogonal to everything else.

Sounds good! I'll move this out of core into its own top-level `history` module.